workflows/differentialabundance.nf

```diff
@@ -168,280 +48,460 @@ include { DIFFERENTIAL_FUNCTIONAL_ENRICHMENT                } from '../subworkfl
+
+    ch_control_features = ch_paramsets
+        .map { paramset -> [ paramset, paramset.control_features ? file(paramset.control_features, checkIfExists: true) : [] ] }
+
+    ch_gene_sets = ch_paramsets
+        .map { paramset -> [ paramset, paramset.gene_sets_files ? paramset.gene_sets_files.split(",").collect { file(it, checkIfExists: true) } : [] ] }
+
+    // ========================================================================
+    // Handle contrasts
+    // ========================================================================
+
+    // Create contrasts channels
+    ch_contrasts_file = ch_paramsets
+        .map { paramset ->
+            [ paramset, file(paramset.contrasts_yml ?: paramset.contrasts, checkIfExists: true) ]
+        }
+
+    ch_contrasts_file_with_extension = ch_contrasts_file
+        .map {
+            paramset, file -> [paramset, file, file.extension]
+        }
+
+    ch_contrast_variables_input = ch_contrasts_file_with_extension
+        .branch{ paramset, file, extension ->
+            yml: extension == 'yml' || extension == 'yaml'
+            csv: extension == 'csv'
+            tsv: extension == 'tsv'
+        }
+
+    ch_contrasts_variables_from_yml = ch_contrast_variables_input.yml
+        .map { paramset, yaml_file, ext ->
+            def yaml_data = new groovy.yaml.YamlSlurper().parse(yaml_file)
+            yaml_data.contrasts.collect { contrast ->
+                tuple(paramset, contrast.comparison[0])
             }
-            .unique()
-    }
+        }
+        .map{it[0]}
+        .unique() // Uniquify to keep each contrast variable only once (in case it exists in multiple lines for blocking etc.)
+
+    ch_contrasts_variables_from_other = ch_contrast_variables_input.csv.splitCsv(header:true)
+        .mix(ch_contrast_variables_input.tsv.splitCsv(header:true, sep:'\t'))
+        .map { paramset, row, ext ->
+            [paramset, row.variable]
+        }
+        .unique()
 
+    ch_contrast_variables = ch_contrasts_variables_from_yml
+        .mix(ch_contrasts_variables_from_other)
+
+    // ========================================================================
+    // Data type specific preprocessing
+    // ========================================================================
+
+    //
+    // 1. deal with affy array data
     // If we have affy array data in the form of CEL files we'll be deriving
     // matrix and annotation from them
+    //
 
-    if (params.study_type == 'affy_array'){
+    // Uncompress the CEL files archive
 
-        // Uncompress the CEL files archive
+    ch_untar_input = prepareModuleInput(ch_celfiles)
 
-        UNTAR ( ch_celfiles )
+    UNTAR ( ch_untar_input )
 
-        ch_affy_input = ch_input
-            .join(UNTAR.out.untar)
+    ch_untar_out = prepareModuleOutput(UNTAR.out.untar, ch_paramsets_by_name)
```

Comment:
this would also need a very good documentation/warning, otherwise I would forget to add these two functions if I want to add a new module